orangecontrib/text/keywords/embedding.py

```diff
@@ -0,0 +1,113 @@
+from collections import defaultdict
+from operator import itemgetter
+from typing import List, Tuple, Dict, Set, Callable
+
+import numpy as np
+from Orange.util import dummy_callback, wrap_callback
+
+from orangecontrib.text import Corpus
+from orangecontrib.text.vectorization import document_embedder
+from orangecontrib.text.vectorization.document_embedder import DocumentEmbedder
+from sklearn.metrics.pairwise import cosine_similarity
+
+
+EMBEDDING_LANGUAGE_MAPPING = document_embedder.LANGS_TO_ISO
+
+
+def _embedd_tokens(
+    tokens: List[List[str]], language: str, progress_callback: Callable
+) -> Tuple[np.ndarray, np.ndarray, Dict[str, Set[int]]]:
+    """
+    Embedd document and words and create a mapping dictionary between words and
+    documents
+    """
+    # extract words
+    word2doc = defaultdict(set)
+    for i, doc_tokens in enumerate(tokens):
+        for t in doc_tokens:
+            word2doc[t].add(i)
+    words = list(word2doc.keys())
+
+    # TODO: currently embedding report success unify them to report progress float
+    ticks = iter(np.linspace(0, 1, len(tokens) + len(words)))
+
+    def emb_cb(sucess: bool):
+        if sucess:
+            progress_callback(next(ticks))
+
+    # embedd documents
+    embedder = DocumentEmbedder(language=language)
+    doc_embs = np.array(embedder(tokens, emb_cb))
+
+    # embedd words
+    word_embs = np.array(embedder([[w] for w in words], emb_cb))
+
+    return doc_embs, word_embs, word2doc
+
+
+def cos_dist(x: np.ndarray, y: np.ndarray) -> np.ndarray:
+    return 1 - cosine_similarity(x, y)
+
+
+def embedding_keywords(
+    tokens: List[List[str]],
```

Comment:
This is actually not true, because Corpus.tokens is a np.ndarray. Therefore, the `_embedd_tokens` crashes. The question is: where should this be handled? In Corpus (store_tokens), OWKeywords, embedding_keywords, _embedd_tokens?

Reply:
You are right. I think that tokens are `np.ndarray` that it supports all kind of selections that Table support e.g. `tokens[[1, 4, 8]]`. I just fixed a method that it can work with both np.ndarray and List.